pkg/sql/queuefeed/reader.go

```diff
@@ -218,9 +210,26 @@ func (r *Reader) GetRows(ctx context.Context, limit int) ([]tree.Datums, error)
 
 	if len(r.mu.buf) == 0 {
 		fmt.Printf("GetRows called with empty buf. waiting for pushedWakeup\n")
-		for ctx.Err() == nil && len(r.mu.buf) == 0 {
+		// shut down the reader if this ctx (which is distinct from the goro ctx) is canceled
+		done := make(chan struct{})
+		defer close(done)
+		go func() {
```

Comment:
You can avoid creating this goroutine by using context.AfterFunc.

defer context.AfterFunc(ctx, func() { r.cancel(errors.Wrapf(ctx.Err(), "GetRows canceled")) })()